### DISCUSSION TITLE
Spurious structure for sparse matrices?

### DISCUSSION THREAD

**Discussion by deklanw** [lmcinnes/umap#597] [Q&A]
I was playing with UMAP, first with a dense dataset and then later with a sparse dataset.

Here's the result I got with all defaults for the dense dataset

![image](https://user-images.githubusercontent.com/11601778/108404000-56a87300-71ed-11eb-97a7-b1a63a15fbe1.png)

Looked interesting. To sanity check it I just used `np.random.rand` to generate data of the same shape,

![image](https://user-images.githubusercontent.com/11601778/108404056-6a53d980-71ed-11eb-9ad3-3220a4d15d72.png)

Looks like a random mess. So far so good.

I tried the same thing with a sparse binary real-world dataset. Result:

![image](https://user-images.githubusercontent.com/11601778/108404192-93746a00-71ed-11eb-8c45-f9b45a7fbd03.png)

And, then sanity checked with a random sparse binary matrix with the same density and shape as my original:

```python

R_density = R.nnz / np.multiply(*R.shape)
random_binary = scipy.sparse.random(*R.shape, density=R_density, format='csr')

# replace nonzero with 1
random_binary.data[:] = 1
```

Result:

![image](https://user-images.githubusercontent.com/11601778/108405006-8f951780-71ee-11eb-9440-8fa012c4b611.png)

This doesn't look like a random mess. I removed the binary requirement and got the same result. Anyone know what would cause this?

**Comment by lmcinnes** [comment]
The only real difference is in the nearest neighbor search which did have some minor bugs in the sparse computations at one time. That is all farmed out to PyNNDescent these days and those bugs have been fixed there, but perhaps there is something odd outstanding? I am not aware of any reason why this should occur other than, perhaps, the sparse rp-trees giving some weird trouble.
